File: scripts/python/part-iv/summarize_landings.py
"""Valida y resume el caso sintético con la biblioteca estándar de Python."""

from __future__ import annotations

import argparse
import csv
from collections import defaultdict
from datetime import date
from pathlib import Path

REQUIRED = {"date", "port", "fleet", "species", "landings_t", "source_id", "quality_flag"}
VALID_FLAGS = {"ok", "review", "invalid"}


def read_landings(path: Path) -> list[dict[str, str]]:
    with path.open(encoding="utf-8", newline="") as source:
        reader = csv.DictReader(source)
        if set(reader.fieldnames or []) != REQUIRED:
            raise ValueError(f"Esquema inesperado: {reader.fieldnames}")
        return list(reader)


def validate(rows: list[dict[str, str]]) -> list[str]:
    errors: list[str] = []
    seen: set[str] = set()
    for number, row in enumerate(rows, start=2):
        try:
            date.fromisoformat(row["date"])
            value = float(row["landings_t"])
        except ValueError as exc:
            errors.append(f"fila {number}: tipo inválido ({exc})")
            continue
        if row["source_id"] in seen:
            errors.append(f"fila {number}: source_id duplicado")
        seen.add(row["source_id"])
        if row["quality_flag"] not in VALID_FLAGS:
            errors.append(f"fila {number}: quality_flag desconocido")
        if value < 0 and row["quality_flag"] != "invalid":
            errors.append(f"fila {number}: valor negativo sin marca invalid")
    return errors


def summarize(rows: list[dict[str, str]]) -> dict[str, float]:
    totals: defaultdict[str, float] = defaultdict(float)
    for row in rows:
        if row["quality_flag"] != "invalid":
            totals[row["species"]] += float(row["landings_t"])
    return dict(sorted(totals.items()))


def write_summary(summary: dict[str, float], path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8", newline="") as target:
        writer = csv.writer(target)
        writer.writerow(["species", "landings_t"])
        writer.writerows((species, f"{value:.1f}") for species, value in summary.items())


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("input", type=Path)
    parser.add_argument("output", type=Path)
    args = parser.parse_args()
    try:
        rows = read_landings(args.input)
    except (OSError, ValueError) as exc:
        raise SystemExit(f"Validación fallida: {exc}") from exc
    errors = validate(rows)
    if errors:
        raise SystemExit("Validación fallida:\n- " + "\n- ".join(errors))
    summary = summarize(rows)
    write_summary(summary, args.output)
    excluded = sum(row["quality_flag"] == "invalid" for row in rows)
    print(
        f"Validación superada: {len(rows)} registros leídos, "
        f"{excluded} excluidos y {len(summary)} especies resumidas."
    )
    print(f"Salida escrita: {args.output}")


if __name__ == "__main__":
    main()

File: scripts/r/part-iv/summarize_landings.R
#!/usr/bin/env Rscript

# Valida y resume el caso sintético usando únicamente R base.
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 2L) {
  stop("Uso: Rscript summarize_landings.R <entrada.csv> <salida.csv>")
}

required <- c("date", "port", "fleet", "species", "landings_t", "source_id", "quality_flag")
input <- read.csv(args[[1]], stringsAsFactors = FALSE, check.names = FALSE)

if (!setequal(names(input), required)) stop("El esquema de columnas no coincide")
if (anyDuplicated(input$source_id)) stop("source_id contiene duplicados")
if (any(is.na(as.Date(input$date)))) stop("date contiene valores inválidos")
numeric_landings <- suppressWarnings(as.numeric(input$landings_t))
if (any(is.na(numeric_landings))) stop("landings_t contiene valores no numéricos")
input$landings_t <- numeric_landings
if (any(!input$quality_flag %in% c("ok", "review", "invalid"))) stop("quality_flag desconocido")
if (any(input$landings_t < 0 & input$quality_flag != "invalid")) {
  stop("Un valor negativo carece de la marca invalid")
}

valid <- input[input$quality_flag != "invalid", ]
summary <- aggregate(landings_t ~ species, data = valid, FUN = sum)
summary <- summary[order(summary$species), ]
dir.create(dirname(args[[2]]), recursive = TRUE, showWarnings = FALSE)
write.csv(summary, args[[2]], row.names = FALSE, quote = FALSE)
message(
  sprintf(
    "Validación superada: %d registros leídos, %d excluidos y %d especies resumidas.",
    nrow(input), sum(input$quality_flag == "invalid"), nrow(summary)
  )
)
message(sprintf("Salida escrita: %s", args[[2]]))
